pkg/appgw/backendhttpsettings.go

```diff
@@ -73,148 +73,139 @@ func (c *appGwConfigBuilder) getBackendsAndSettingsMap(cbCtx *ConfigBuilderConte
 			}
+			resolvedBackendPorts[pair] = nil
+			continue
 		}
 
-		if len(serviceBackendPairs) > 1 {
+		// if target port is an int, use it as backend port
+		if servicePort.TargetPort.Type == intstr.Int {
+			// port is defined as port number
+			pair := serviceBackendPortPair{
+				ServicePort: Port(servicePort.Port),
+				BackendPort: Port(servicePort.TargetPort.IntVal),
+			}
+			resolvedBackendPorts[pair] = nil
+			continue
+		}
 
-			// more than one possible backend port exposed through ingress
-			e := controllererrors.NewErrorf(
-				controllererrors.ErrorMultipleServiceBackendPortBinding,
-				"service:port [%s:%s] has more than one service-backend port binding which is not an ideal scenario, choosing the smallest service-backend port %d. Ports found %s.",
-				backendID.serviceKey(), backendID.Backend.ServicePort.String(), backendport, strings.Join(ports, ","),
-			)
+		// if target port is port name, then resolve the port number for the port name
+		// k8s matches service port name against endpoints port name retrieved by passing backendID service key to endpoint api.
+		klog.V(5).Infof("resolving port name '%s' for service '%s' and service port '%s' for Ingress '%s'", servicePort.Name, backendID.serviceKey(), backendID.Backend.ServicePort.String(), backendID.Ingress.Name)
+		targetPortsResolved := c.resolvePortName(servicePort.Name, &backendID)
+		for targetPort := range targetPortsResolved {
+			pair := serviceBackendPortPair{
+				ServicePort: Port(servicePort.Port),
+				BackendPort: Port(targetPort),
+			}
+			resolvedBackendPorts[pair] = nil
+		}
+	}
 
-			c.recorder.Event(backendID.Ingress, v1.EventTypeWarning, events.ReasonPortResolutionError, e.Error())
-			klog.Errorf(e.Error())
+	if len(resolvedBackendPorts) == 0 {
+		e = controllererrors.NewErrorf(
+			controllererrors.ErrorUnableToResolveBackendPortFromServicePort,
+			"No port matched %s",
+			backendID.serviceKey())
+
+		// if service port is an int, use that as backend port
+		if backendID.Backend.ServicePort.Type == intstr.Int {
+			return Port(backendID.Backend.ServicePort.IntVal), e
 		}
 
-		finalServiceBackendPairMap[backendID] = uniquePair
-		httpSettings := c.generateHTTPSettings(backendID, uniquePair.BackendPort, cbCtx)
-		klog.V(5).Infof("Created backend http settings %s for ingress %s/%s and service %s", *httpSettings.Name, backendID.Ingress.Namespace, backendID.Ingress.Name, backendID.serviceKey())
-		httpSettingsCollection[*httpSettings.Name] = httpSettings
-		backendHTTPSettingsMap[backendID] = &httpSettings
+		// returning port 80 as a last resort
+		return Port(80), e
 	}
 
-	httpSettings := make([]n.ApplicationGatewayBackendHTTPSettings, 0, len(httpSettingsCollection))
-	for _, backend := range httpSettingsCollection {
-		httpSettings = append(httpSettings, backend)
+	backendPort := Port(65536)
+	var ports []string
+	for k := range resolvedBackendPorts {
+		ports = append(ports, string(k.BackendPort))
+		if k.BackendPort <= backendPort {
```

Comment:
if k.BackendPort > 65536, then we need error out as well, otherwise 65536 will be used as default in the following